### PR TITLE
c++ -> closing the farm (gold) 

### PR DESCRIPTION
**Pull request by ryanchou-dev**
blahblahblah

---
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

File: solutions/usaco-646.mdx [hunk 1]
```diff
@@ -26,7 +26,96 @@ Note: The video solution might not be the same as other solutions. Code in C++.
...
+				if (dsu.unite(j, rev[i])) {
+					cc--;
+				}
+			}
+		}
+
+		if (cc == 1) {
+			ans.push_back("YES");
+		} else {
+			ans.push_back("NO");
```
Comment (SansPapyrus683) [inline solutions/usaco-646.mdx:108]: ternary?

File: solutions/usaco-646.mdx [hunk 2]
```diff
@@ -26,7 +26,96 @@ Note: The video solution might not be the same as other solutions. Code in C++.
...
+		}
+	}
+	reverse(ans.begin(), ans.end());
+
+	for (auto i : ans) {
+		cout << i << endl;
```
Comment (SansPapyrus683) [inline solutions/usaco-646.mdx:114]: i know i said i like endl but in a for loop i've decided that \n is better

File: solutions/usaco-646.mdx [hunk 3]
```diff
@@ -26,7 +26,96 @@ Note: The video solution might not be the same as other solutions. Code in C++.
...
+
+		if (cc == 1) {
+			ans.push_back("YES");
+		} else {
+			ans.push_back("NO");
+		}
+	}
+	reverse(ans.begin(), ans.end());
+
+	for (auto i : ans) {
```
Comment (SansPapyrus683) [inline solutions/usaco-646.mdx:113]:
```suggestion
	for (const string& i : ans) {
```

File: solutions/usaco-646.mdx [hunk 4]
```diff
@@ -26,7 +26,96 @@ Note: The video solution might not be the same as other solutions. Code in C++.
...
+	DSU(int N) { e = vector<int>(N, -1); }
+
+	// get representive component (uses path compression)
+	int get(int x) { return e[x] < 0 ? x : e[x] = get(e[x]); }
+
+	bool same_set(int a, int b) { return get(a) == get(b); }
+
+	int size(int x) { return -e[get(x)]; }
+
+	bool unite(int x, int y) {	// union by size
```
Comment (SansPapyrus683) [inline solutions/usaco-646.mdx:49]:
```suggestion
	// union by size
	bool unite(int x, int y) {
```

File: solutions/usaco-646.mdx [hunk 5]
```diff
@@ -26,7 +26,96 @@ Note: The video solution might not be the same as other solutions. Code in C++.
...
+
+	int n, m;
+	cin >> n >> m;
+
+	vector<vector<int>> adj(n);
+	for (int i = 0; i < m; i++) {
+		int u, v;
+		cin >> u >> v;
+		--u;
+		--v;
```
Comment (SansPapyrus683) [inline solutions/usaco-646.mdx:72]:
```suggestion
		u--;
		v--;

File: solutions/usaco-646.mdx [hunk 6]
```diff
@@ -26,7 +26,96 @@ Note: The video solution might not be the same as other solutions. Code in C++.
 **Time Complexity:** $\mathcal{O}(V \cdot E \cdot \alpha(n)))$
 
 <LanguageSection>
+<CPPSection>
 
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+using ll = long long;
```
Comment (SansPapyrus683) [inline solutions/usaco-646.mdx:35]: i don't see you using this?

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-646.mdx (hunk 2 above)
Comment (ryanchou-dev) [reply solutions/usaco-646.mdx:115]: burh

**Review by SansPapyrus683 — APPROVED** [review]

File: solutions/usaco-646.mdx [hunk 7]
```diff
@@ -26,7 +26,89 @@ Note: The video solution might not be the same as other solutions. Code in C++.
...
+<CPPSection>
 
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+//BeginCodeSnip{DSU}
+struct DSU {
+	vector<int> e;
+	DSU(int N) { e = vector<int>(N, -1); }
```
Comment (ChuangSheep) [inline solutions/usaco-646.mdx:38]:
```suggestion
	DSU(int N): e(N, -1) { }
```
optional, but its possible to use initializer lists

**Review by ChuangSheep — APPROVED** [review]
nice

**Pull request merged**